kernel: timeout: do not active time slicing if idle thread ready

wentongwu opened the pull request:
In one time slicing interval, there is a swap from the only working thread to idle thread, so the _current will be _idle_thread, when the timer interrupt comes later, in function z_time_slice, _current_cpu->slice_ticks will not be zero, so in next_timeout it will always return the current value of _current_cpu->slice_ticks and that will cause time slicing wake up system periodically,  the value of _current_cpu->slice_ticks in the idle thread condition and resume it appropriately, so that system can stay low power state longer time.

Fixes: #17368.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

pizi-nordic commented on kernel/timeout.c:
```diff
@@ -73,7 +73,17 @@ static s32_t next_timeout(void)
 
 #ifdef CONFIG_TIMESLICING
 	if (_current_cpu->slice_ticks && _current_cpu->slice_ticks < ret) {
+#ifndef CONFIG_SMP
```
Why we cannot just set slice_ticks = 0 for idle thread?

wentongwu replied:
> Why we cannot just set slice_ticks = 0 for idle thread?

@pizi-nordic thanks for the review, yes, changing time slicing related variables is more reasonable. But my trigger point is in timeout.c, I don't want to modify that variables outside sched.c. Maybe we should add more interfaces in sched.c to enable/disable time slicing for kernel side without changing slice_time. Thanks again.

pizi-nordic replied:
My fault: I mistaken _current_cpu with _current :).
What about changing condition to:
```
if ((_current_cpu->slice_ticks && _current_cpu->slice_ticks < ret && sliceable(_current))
```
?

wentongwu replied:
@pizi-nordic thanks. I think sliceable(_idle_thread) equals to my first condition (_current != _idle_thread), but we have to consider another case, _current == _idle_thread && _kernel.ready_q.cache != _idle_thread, this is because when _current is _idle_thread and a external interrupt or a timer interrupt wakeup a real thread, we have to resume the time slicing function timely. And I think my changes "(_current != _idle_thread || _kernel.ready_q.cache != _idle_thread) " can cover above concerns. Thanks

wentongwu replied:
Also, most likely, this change just add this "_current != _idle_thread" overhead at run time.

pizi-nordic replied:
I do not like the way time slicing is implemented as we have to workaround bugs created by poor design.

The is reset_time_slice() called when ready cache is updated and this sets the timeout resuming slicing operation. But then timer interrupt might fire and overwrite this information. I am not sure if this is the scenario you have in mind, but I see why you have to check ready queue.

wentongwu replied:
> The is reset_time_slice() called when ready cache is updated and this sets the timeout resuming slicing operation. But then timer interrupt might fire and overwrite this information. I am not sure if this is the scenario you have in mind, but I see why you have to check ready queue.

yes, that's my consideration. Thanks.

andyross reviewed (changes requested):
This seems like the wrong place for the fix.  This is already supposed to be handled in sched.c:: there is a sliceable() predicate which already checks is_idle(), and is already used in the z_time_slice() hook when deciding whether to register the new slice ticks.

AFAICT the bug is in the SWAP_NONATOMIC workaround at the top of that function -- it's calling reset_time_slice() without checking sliceable() first.

andyross commented:
Can you try this and see if it fixes the "idle is timeslicing" bug?

```patch
diff --git a/kernel/sched.c b/kernel/sched.c
index be1b5dec2b..23bbd6d596 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -283,7 +283,7 @@ static inline int sliceable(struct k_thread *t)
 void z_time_slice(int ticks)
 {
 #ifdef CONFIG_SWAP_NONATOMIC
-       if (pending_current == _current) {
+       if (pending_current == _current && sliceable(_current)) {
                reset_time_slice();
                return;
        }
```

wentongwu commented:
> This seems like the wrong place for the fix. This is already supposed to be handled in sched.c:: there is a sliceable() predicate which already checks is_idle(), and is already used in the z_time_slice() hook when deciding whether to register the new slice ticks.
> 
> AFAICT the bug is in the SWAP_NONATOMIC workaround at the top of that function -- it's calling reset_time_slice() without checking sliceable() first.

@andyross  Thanks for the review, yes, I will try the patch. But consider one case, in one time slicing interval, there is a swap from the only working thread to idle thread, so the _current will be _idle_thread, when the timer interrupt comes later, in function z_time_slice, _current_cpu->slice_ticks will not be zero, so in next_timeout it will always return the current value of _current_cpu->slice_ticks.

wentongwu commented:
> Can you try this and see if it fixes the "idle is timeslicing" bug?
> 
> ```diff
> diff --git a/kernel/sched.c b/kernel/sched.c
> index be1b5dec2b..23bbd6d596 100644
> --- a/kernel/sched.c
> +++ b/kernel/sched.c
> @@ -283,7 +283,7 @@ static inline int sliceable(struct k_thread *t)
>  void z_time_slice(int ticks)
>  {
>  #ifdef CONFIG_SWAP_NONATOMIC
> -       if (pending_current == _current) {
> +       if (pending_current == _current && sliceable(_current)) {
>                 reset_time_slice();
>                 return;
>         }
> ```

@andyross  I tried the patch, and hit once the swap non-atomic case, but also get the case as last comment, I'm sorry for the not clearly root cause words.
And it seems the enable/disable of time slicing is controlled by slice_time variable, but in next_timeout _current_cpu->slice_ticks is also used for the control, I'm not familiar with the design, so made the patch to bypass the value of _current_cpu->slice_ticks in the idle thread condition and resume it appropriately.

andyross reviewed (changes requested):
Just getting back to this.  I'm still not understanding the problem here.  It's true that if you z_swap() to the idle thread that the time slice state doesn't get informed and will deliver the interrupt it originally scheduled.  But... so what?  It's a cost of one timer interrupt per idle entry.

Is that what you're trying to fix?  The one timer interrupt?  Can you post an exerciser for this bug?  I'm not understanding.

andyross commented:
Oh wait... I think I see.  You're saying that the time slicing keeps re-registering because next_timeout() tells the kernel that there is one pending, even though the time slice handling is ignoring it.  Yes, that's a bug.

Andy this is definitely fixing the wrong spot.  :) Don't put code into every call to next_timeout().  Put code into z_time_slice() to zero the tick count, which runs much less often.  I'm guessing this might fix your bug?  (Again, I'm not sure how you're exercising this to test against, so this is written blindly):

```patch
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -296,6 +296,8 @@ void z_time_slice(int ticks)
                } else {
                        _current_cpu->slice_ticks -= ticks;
                }
+       } else {
+               _current_cpu->slice_ticks = 0;
        }
 }
 #else
```

wentongwu commented:
> Oh wait... I think I see. You're saying that the time slicing keeps re-registering because next_timeout() tells the kernel that there is one pending, even though the time slice handling is ignoring it. Yes, that's a bug.
> 
> And this is definitely fixing the wrong spot. :) Don't put code into every call to next_timeout(). Put code into z_time_slice() to zero the tick count, which runs much less often. I'm guessing this might fix your bug? (Again, I'm not sure how you're exercising this to test against, so this is written blindly):
> 
> ```diff
> --- a/kernel/sched.c
> +++ b/kernel/sched.c
> @@ -296,6 +296,8 @@ void z_time_slice(int ticks)
>                 } else {
>                         _current_cpu->slice_ticks -= ticks;
>                 }
> +       } else {
> +               _current_cpu->slice_ticks = 0;
>         }
>  }
>  #else
> ```

ok, it's more reasonable than the PR's change. I will test this patch.
And the test case is the main thread(only thread in the system except idle thread) just do sleeping when it get scheduled.

wentongwu commented:
@andyross  I tested the changes as you suggested, it also works well. Thanks

wentongwu commented:
@pizi-nordic as Andy's suggestion, there is a more reasonable patch for this issue, could you please re-visit this PR? thanks

wentongwu committed:
kernel: timeout: do not active time slicing if idle thread ready

zero slice_ticks when can't time slice so that next_timeout will
ignore slice_ticks of _current_cpu and system can stay low power
state longer time.

Fixes: #17368.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

andyross reviewed (approved):
Nice.  I'm actually amazed that worked without modification.  Usually scheduler behavior is a little more subtle.
